### ISSUE TITLE
Question about the loss computation 

### ISSUE THREAD

**Issue by edwardpwtsoi** [tfzhou/ContrastiveSeg#20]
https://github.com/tfzhou/ContrastiveSeg/blob/2ab84d8ec679adc7f7be1853c8684b44bf899273/lib/loss/loss_contrast.py#L106

Would you mind explain why is it needed to subtract max of the inner product from each of the inner product per anchor?

**Comment by HenryPengZou** [comment]
Same question here

**Comment by eezywu** [comment]
To avoid the numerical overflow of softmax function. It is equivalent whether to subtract `logits_max.detach()`.

**Comment by edwardpwtsoi** [comment]
> To avoid the numerical overflow of softmax function. It is equivalent whether to subtract `logits_max.detach()`.

isn't the feature vector already l2 normalised? so this subtraction would just shift the range of the inner product from [0, 1) to [-1, 0)

**Comment by tfzhou** [comment]
Yes, it is to avoid the numerical overflow to subtract ``logits_max.detach()``

**Comment by edwardpwtsoi** [comment]
Thx! @eezywu and @tfzhou, I got it now
